### PR TITLE
share validator keys in EpochRef

### PR DESCRIPTION
**Pull request by arnetheduck**

### CODE REVIEW

File: beacon_chain/block_pools/chain_dag.nim [hunk 1]
```diff
@@ -77,7 +77,17 @@ proc init*(T: type EpochRef, state: BeaconState, cache: var StateCache): T =
       epochRef.beacon_proposers[i] =
         some((idx.get(), state.validators[idx.get].pubkey))
 
-  epochRef.validator_keys = mapIt(state.validators.toSeq, it.pubkey)
+  if prevEpoch != nil and
+      (prevEpoch.validator_key_store[0] == hash_tree_root(state.validators)):
+    # Validator sets typically don't change between epochs - a more efficient
```
Comment (tersec) [inline beacon_chain/block_pools/chain_dag.nim:82]: I'm not sure that this is true in Medalla at the moment -- there are still activation queues of days for new validators. But yes, this is probably typically true.

**Comment by tersec** [comment]
It's a bit odd though,  it runs in CI finalization at least about as much as `devel` typically does, but running it locally, switching between `devel` and this branch, on this branch I get:
```Last 20 lines of local_testnet_data/log3.txt:
{"lvl":"INF","ts":"2020-08-11 20:43:45.539+02:00","msg":"No attestations matching the slot range","topics":"attpool","tid":199944,"file":"attestation_pool.nim:234","attestationSlot":2,"startingSlot":4}
{"lvl":"INF","ts":"2020-08-11 20:43:45.539+02:00","msg":"No attestations matching the slot range","topics":"attpool","tid":199944,"file":"attestation_pool.nim:234","attestationSlot":3,"startingSlot":4}
{"lvl":"DBG","ts":"2020-08-11 20:43:45.587+02:00","msg":"Less peers in routing table than maximum requeTraceback (most recent call last, using override)
/home/user/nim-beacon-chain/beacon_chain/mainchain_monitor.nim(543) main
/home/user/nim-beacon-chain/beacon_chain/mainchain_monitor.nim(536) NimMain
/home/user/nim-beacon-chain/beacon_chain/beacon_node.nim(1232) main
/home/user/nim-beacon-chain/beacon_chain/beacon_node.nim(926) start
/home/user/nim-beacon-chain/vendor/nim-chronicles/chronicles.nim(329) run
/home/user/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(407) reportUnhandledError
/home/user/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(358) reportUnhandledErrorAux
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
sted","tid":199944,"file":"routing_table.nim:362","requested":76,"present":3}
```

and
```
======
Last 20 lines of local_testnet_data/log2.txt:
/home/user/nim-beacon-chain/beacon_chain/beacon_node.nim(1232) main
/home/user/nim-beacon-chain/beacon_chain/beacon_node.nim(926) start
/home/user/nim-beacon-chain/vendor/nim-chronicles/chronicles.nim(329) run
/home/user/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(407) reportUnhandledError
/home/user/nim-beacon-chain/vendor/nimbus-build-system/vendor/Nim/lib/system/excpt.nim(358) reportUnhandledErrorAux
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
```

At at 20:43:45, resulting in by 20:45, log1 and log0 showing:
```
"lvl":"FAT","ts":"2020-08-11 20:45:17.498+02:00","msg":"Missing too many attesters from previous epoch in verifyFinalization mode","topics":"consens","tid":199804,"file":"state_transition_epoch.nim:146","missing_all_validators":[65,2,67,4,6,7,72,73,10,13,78,15,80,81,18,84,85,25,30,95,96,0,98,100,109,110,47,111,113,51,53,54,117,119,59,123,61,62],"pcs":"process_justification_and_finalization"
```

It's not obvious to me why that failure mode would correspond to the changes in this PR, so it's a sort of tentative approval. In particular, something similar, but not necessarily identical, happens on `devel` as well, and I don't yet have the sample size to distinguish between the distributions.

So this seems probably okay: I can barely get either to finalize locally, for reasons seemingly unrelated to this PR.

File: beacon_chain/block_pools/chain_dag.nim (hunk 1 above)
Comment (arnetheduck) [reply beacon_chain/block_pools/chain_dag.nim:82]: even medalla, the blocks don't have deposits in them mostly: https://medalla.beaconcha.in/blocks

**Comment by tersec** [comment]
Ah, got a stack trace which makes more sense:
```
Last 20 lines of local_testnet_data/log2.txt:
beacon_chain/block_pools/chain_dag.nim(202) getEpochCache
beacon_chain/block_pools/chain_dag.nim(174) getEpochInfo
vendor/nimbus-build-system/vendor/Nim/lib/system.nim(2522) []
vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal
Error: unhandled exception: index out of bounds, the container is empty [IndexError]
```
So `blck.epochsInfo.filterIt(it.epoch < state_epoch)` is empty but `blck.epochsInfo.len > 0`.

**Comment by arnetheduck** [comment]
> is empty but

I force-pushed that one away before merging - it was an actual bug caught by the finalization test ;)

**Comment by tersec** [comment]
Meanwhile, testing with
```diff --git a/beacon_chain/block_pools/chain_dag.nim b/beacon_chain/block_pools/chain_dag.nim
index 90a33a4..d6c1c00 100644
--- a/beacon_chain/block_pools/chain_dag.nim
+++ b/beacon_chain/block_pools/chain_dag.nim
@@ -170,7 +170,7 @@ proc getEpochInfo*(blck: BlockRef, state: BeaconState, cache: var StateCache): E
 
     var prevEpochRefs = blck.epochsInfo.filterIt(it.epoch < state_epoch)
     var prevEpochRef: EpochRef = nil # nil ok
-    if blck.epochsInfo.len > 0:
+    if prevEpochRefs.len > 0:
       prevEpochRef = prevEpochRefs[^1]
     elif state_epoch > 0:
       let parent = blck.atEpochEnd((state_epoch - 1))
```

Which seems like the most obvious fix, also for the other PR, since you rebased there.

**Pull request merged**